server/src/metakb/load_data.py

```diff
@@ -202,76 +202,149 @@ def _add_therapy(tx: ManagedTransaction, therapy_in: dict) -> None:
+        end: $allele.location.end,
+        refget_accession: $allele.location.sequenceReference.refgetAccession,
+        sequence: $allele.location.sequence
+    }
+    MERGE (allele) -[:HAS_LOCATION]-> (sl)
+
+    // handle different kinds of state objects
+    FOREACH (_ IN CASE WHEN $allele.state.type = 'LiteralSequenceExpression' THEN [1] ELSE [] END |
+        MERGE (lse:SequenceExpression:LiteralSequenceExpression { sequence: $allele.state.sequence })
+        MERGE (allele)-[:HAS_STATE]->(lse)
+    )
+    FOREACH (_ IN CASE WHEN $allele.state.type = 'ReferenceLengthExpression' THEN [1] ELSE [] END |
+        MERGE (rle:SequenceExpression:ReferenceLengthExpression {
+            length: $allele.state.length,
+            repeat_subunit_length: $allele.state.repeatSubunitLength,
+            sequence: $allele.state.sequence
+        })
+        MERGE (allele)-[:HAS_STATE]->(rle)
+    )
 
-    query = f"""
-    MERGE (v:{v["type"]}:Variation {{ id: '{v["id"]}' }})
-    ON CREATE SET {v_keys}
+    WITH cv
+        UNWIND $members as m
+        MERGE (member_allele:Variation:MolecularVariation:Allele { id: m.id })
+        ON CREATE SET member_allele += {
+            name: m.name,
+            digest: m.digest,
+            expression_hgvs_g: m.expression_hgvs_g,
+            expression_hgvs_c: m.expression_hgvs_c,
+            expression_hgvs_p: m.expression_hgvs_p
+        }
+        MERGE (cv) -[:HAS_MEMBER]-> (member_allele)
+        MERGE (member_sl:Location:SequenceLocation { id: m.location.id })
+        ON CREATE SET member_sl += {
+            digest: m.location.digest,
+            start: m.location.start,
+            end:  m.location.end,
+            refget_accession: m.location.sequenceReference.refgetAccession,
+            sequence: m.location.sequence
+        }
+        MERGE (member_allele) -[:HAS_LOCATION] -> (member_sl)
+
+        // handle different kinds of state objects
+        FOREACH (_ IN CASE WHEN m.state.type = 'LiteralSequenceExpression' THEN [1] ELSE [] END |
+            MERGE (member_lse:SequenceExpression:LiteralSequenceExpression { sequence: m.state.sequence })
+            MERGE (member_allele)-[:HAS_STATE]->(member_lse)
+        )
+        FOREACH (_ IN CASE WHEN m.state.type = 'ReferenceLengthExpression' THEN [1] ELSE [] END |
+            MERGE (member_rle:SequenceExpression:ReferenceLengthExpression {
+                length: m.state.length,
+                repeat_subunit_length: m.state.repeatSubunitLength,
+                sequence: m.state.sequence
+            })
+            MERGE (member_allele)-[:HAS_STATE]->(member_rle)
+        )
     """
 
-    loc = v.get("location")
-    if loc:
-        _add_location(tx, loc)
-        query += f"""
-        MERGE (loc:{loc["type"]}:Location {{ id: '{loc["id"]}' }})
-        MERGE (v) -[:HAS_LOCATION] -> (loc)
-        """
-
-    tx.run(query, **v)
+    # catvars currently support a single constraint
+    constraint = catvar["constraints"][0]
```

Comment:
I know in `add_categorical_variant` below, you check to make sure `constraint` exists/is nonempty - are there other potential entry-points for this function in which we should validate that in this function as well?

Reply:
These objects are basically serialized-then-deserialized pydantic objects, so we do have some guarantees about what's inside them (with the caveat that an evil demon could've slipped in and messed with them during that transition). What I'd like to be doing for these functions is to have them take realized Pydantic classes as input so those guarantees are stricter -- that's part of the big storage arch epic i'm pitching (linked below, described on call today)